HumanControls/HumanControls.ino

```diff
@@ -16,72 +14,136 @@
+boolean hasRobotDataChanged = false;
+boolean hasControllerMessage = false;
+boolean hasControllerDataChanged = false;
+
+uint8_t currentControllerMessage[7];
+uint8_t lastControllerMessage[7];
+uint8_t currentRobotMessage[7];
+uint8_t lastRobotMessage[7];
+
 // Create payload object
 TShirtCannonPayload payload;
 
-HumanControls humanControls(payload, ENCODER_PIN_CLK, ENCODER_PIN_DT, DISPLAY_ADDRESS, DISPLAY_LENGTH, DISPLAY_WIDTH,
-                            ANGLE_INCREMENT, ANGLE_MIN, ANGLE_MAX, PRESSURE_INCREMENT, PRESSURE_MIN, PRESSURE_MAX,
-                            DURATION_INCREMENT, DURATION_MIN, DURATION_MAX, HANG_TIMER_DURATION, DOWN_ARROW,
-                            UP_ARROW, ROBOT_BATTERY_CHAR, CONTROLLER_BATTERY_CHAR, NUM_BUTTONS, ENCODER_PIN_SW,
-                            ENABLE_PIN, PRIME_PIN, FIRE_PIN, JOYSTICK_PIN_VRX, X_DEAD_ZONE_SIZE, JOYSTICK_MAX,
+CommunicationDriver commDriver(RFM95_FREQ, RFM95_TX_POWER, PAYLOAD_LENGTH, RFM95_CS, RFM95_INT, RFM95_RST);
+HumanControls humanControls(payload, ENCODER_PIN_A, ENCODER_PIN_B, ANGLE_INCREMENT, ANGLE_MIN, ANGLE_MAX,
+                            PRESSURE_INCREMENT, PRESSURE_MIN, PRESSURE_MAX, DURATION_INCREMENT, DURATION_MIN,
+                            DURATION_MAX, HANG_TIMER_DURATION, NUM_BUTTONS, ENCODER_PIN_SW, ENABLE_PIN,
+                            PRIME_PIN, FIRE_PIN, JOYSTICK_PIN_VRX, X_DEAD_ZONE_SIZE, JOYSTICK_MAX,
                             JOYSTICK_PIN_VRY, Y_DEAD_ZONE_SIZE);
 
+// Comms core
 void setup()
 {
-    Serial.begin(USB_BAUDRATE);
-    humanControls.init(DOWN_ARROW, UP_ARROW);
+  Serial.begin(USB_BAUDRATE);
+  commDriver.connect(payload);
 }
 
 void loop()
 {
-    humanControls.update();
+  if (hasControllerMessage)
+  {
+    commDriver.sendControllerMessage(currentControllerMessage);
+    hasControllerMessage = false;
+    delay(WAIT_FOR_CONTROLS_DATA_DELAY_MILLIS);
+  }
+  if (hasRobotMessage)
+  {
+    // TODO: log error. Control loop taking too long with robot message
+  }
+  delay(WAIT_FOR_ROBOT_MESSAGE_DELAY_MILLIS);
+  hasRobotMessage = commDriver.recvRobotMessage(currentRobotMessage);
+}
+
+// Controls core
+void setup1()
+{
+  humanControls.init();
+}
+
+void loop1()
+{
+  if (hasRobotMessage)
+  {
+    if (memcmp(currentRobotMessage, lastRobotMessage, PAYLOAD_LENGTH) != 0)
+    {
+      hasRobotDataChanged = true;
+    }
+    memcpy(lastRobotMessage, currentRobotMessage, PAYLOAD_LENGTH);
+    hasRobotMessage = false;
+    if (hasRobotDataChanged)
+    {
+      payload.readMessage(currentRobotMessage, PAYLOAD_LENGTH);
+      hasRobotDataChanged = false;
+    }
+  }
+  humanControls.update(currentControllerMessage);
+  if (!hasControllerMessage)
+  {
+    // memcmp is 1 when no data has changed because of messageIndex
+    if (memcmp(currentControllerMessage, lastControllerMessage, sizeof(currentControllerMessage)) != 1)
+    {
+      hasControllerMessage = true;
+      payload.setMessageIndex((payload.getMessageIndex() + 1) % 32);
```

Comment:
Nothing to do right now, but in the future we might look into refactoring how we handle the payload object so we do not pass payload by reference and set it to member variables. This way it is more obvious that the same payload object is used within member methods across multiple objects.